hivemq-edge/src/main/java/com/hivemq/protocols/ProtocolAdapterWrapper.java

```diff
@@ -88,40 +94,110 @@ public ProtocolAdapterWrapper(
         this.tagManager = tagManager;
     }
 
-    public @NotNull CompletableFuture<Void> start(
-            final boolean writingEnabled, final @NotNull ModuleServices moduleServices) {
-        initStartAttempt();
-        final ProtocolAdapterStartOutputImpl output = new ProtocolAdapterStartOutputImpl();
-        final ProtocolAdapterStartInputImpl input = new ProtocolAdapterStartInputImpl(moduleServices);
-        //Instantly go to started so it has to be stopped first before retrying
-        adapter.start(input, output);
-        final CompletableFuture<Void> startFuture = output.getStartFuture();
-
-        return startFuture.thenCompose(r -> {
-            createAndSubscribeTagConsumer();
-            startPolling(protocolAdapterPollingService, input.moduleServices().eventService());
-            return startWriting(writingEnabled, protocolAdapterWritingService);
-        }).thenApply(r -> {
-            setRuntimeStatus(ProtocolAdapterState.RuntimeStatus.STARTED);
-            return null;
-        });
-
+    public @NotNull CompletableFuture<Boolean> startAsync(
+            final boolean writingEnabled,
+            final @NotNull ModuleServices moduleServices) {
+        final var currentFuture = startFutureRef.get();
+        if(currentFuture == null) {
+            synchronized (startFutureRef) {
+                final var currentFutureDoubleCheck = startFutureRef.get();
+                if(currentFutureDoubleCheck != null) {
+                    return currentFutureDoubleCheck;
+                }
+                initStartAttempt();
+                final ProtocolAdapterStartOutputImpl output = new ProtocolAdapterStartOutputImpl();
+                final ProtocolAdapterStartInputImpl input = new ProtocolAdapterStartInputImpl(moduleServices);
+
+                final CompletableFuture<Boolean> startFuture = new CompletableFuture<>();
+                startFutureRef.set(startFuture);
+
+                CompletableFuture
+                        .supplyAsync(() -> {
+                            adapter.start(input, output);
+                            return output.getStartFuture();
+                        })
+                        .thenCompose(Function.identity())
+                        .handle((ignored, t) -> {
+                            if(t == null) {
+                                createAndSubscribeTagConsumer();
+                                startPolling(protocolAdapterPollingService, input.moduleServices().eventService());
+                                return startWriting(writingEnabled, protocolAdapterWritingService)
+                                        .thenApply(v -> {
+                                            log.info("Successfully started adapter with id {}", adapter.getId());
+                                            setRuntimeStatus(ProtocolAdapterState.RuntimeStatus.STARTED);
+                                            startFutureRef.get().complete(true);
```

Comment:
I wonder this is really safe because it is out of `synchronized (startFutureRef)`. If `startAsync()` was called twice at the same time, what would happen?